File: backend/app/schemas.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, field_validator, model_validator


class AppInfo(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    service: str
    version: str


class RootResponse(AppInfo):
    docs_url: str
    contact_url: str
    environment: str


class HealthResponse(AppInfo):
    status: Literal["ok"]


class VersionResponse(AppInfo):
    public_site_url: str


class LeadCreate(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    name: str
    email: str
    business: str
    problem: str | None = None
    impact: str | None = None
    timeline: str = "Flexible, just exploring options"
    package: str | None = None
    not_sure: bool = False

    @field_validator("name", "business", "timeline")
    @classmethod
    def ensure_non_empty(cls, value: str) -> str:
        if len(value) < 2:
            raise ValueError("Must be at least 2 characters long")
        return value

    @field_validator("problem", "impact")
    @classmethod
    def ensure_non_empty_unless_exploring(cls, value: str | None) -> str | None:
        if value is not None and value != "" and len(value) < 2:
            raise ValueError("Must be at least 2 characters long")
        return value

    @model_validator(mode="after")
    def require_problem_unless_exploring(self) -> "LeadCreate":
        if not self.not_sure:
            for field in ("problem", "impact"):
                value = getattr(self, field)
                if value is None or value == "":
                    raise ValueError(
                        f"{field}: Must be at least 2 characters long"
                    )
        return self

    @field_validator("email")
    @classmethod
    def validate_email(cls, value: str) -> str:
        if "@" not in value or value.startswith("@") or value.endswith("@"):
            raise ValueError("Must be a valid email address")
        if "." not in value.rsplit("@", 1)[-1]:
            raise ValueError("Must be a valid email address")
        return value

    @field_validator("package")
    @classmethod
    def validate_package(cls, value: str | None) -> str | None:
        allowed = {"starter", "business", "advanced"}
        if value is None or value == "":
            return None
        if value not in allowed:
            raise ValueError("Package must be starter, business, or advanced")
        return value


class LeadReceipt(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    status: Literal["received"]
    intake_id: str
    message: str
    next_step: str
    received_at: datetime
    package: str | None = None

File: backend/src/schemas.ts
export interface Env {
  APP_NAME?: string;
  APP_VERSION?: string;
  APP_ENV?: string;
  PUBLIC_SITE_URL?: string;
  CORS_ORIGINS?: string;
  LEAD_EMAIL_TO?: string;
  EMAIL?: SendEmail;
}

interface SendEmail {
  send(msg: EmailMessage): Promise<{ messageId: string }>;
}

interface EmailMessage {
  to: string | string[];
  from: { email: string; name?: string };
  replyTo?: string;
  subject: string;
  html?: string;
  text?: string;
}

const ALLOWED_PACKAGES = new Set(["starter", "business", "advanced"]);

export interface LeadInput {
  name: string;
  email: string;
  business: string;
  problem: string;
  impact: string;
  timeline: string;
  package?: string;
}

export interface ValidationError {
  field: string;
  message: string;
}

export function validateLead(data: Record<string, unknown>): {
  lead?: LeadInput;
  errors?: ValidationError[];
} {
  const errors: ValidationError[] = [];

  const name = stringField(data, "name");
  const email = stringField(data, "email");
  const business = stringField(data, "business");
  const problem = stringField(data, "problem");
  const impact = stringField(data, "impact");
  const timeline = stringField(data, "timeline");
  const pkg = data["package"];

  if (!name || name.length < 2)
    errors.push({ field: "name", message: "Must be at least 2 characters" });
  if (!email || !isValidEmail(email))
    errors.push({ field: "email", message: "Must be a valid email address" });
  if (!business || business.length < 2)
    errors.push({ field: "business", message: "Must be at least 2 characters" });
  if (!problem || problem.length < 2)
    errors.push({ field: "problem", message: "Must be at least 2 characters" });
  if (!impact || impact.length < 2)
    errors.push({ field: "impact", message: "Must be at least 2 characters" });

  if (pkg !== undefined && pkg !== null && pkg !== "" && !ALLOWED_PACKAGES.has(String(pkg))) {
    errors.push({ field: "package", message: "Package must be starter, business, or advanced" });
  }

  if (errors.length > 0) return { errors };

  return {
    lead: {
      name: name!,
      email: email!,
      business: business!,
      problem: problem!,
      impact: impact!,
      timeline: timeline || "Flexible, just exploring options",
      package: pkg ? String(pkg) : undefined,
    },
  };
}

function stringField(data: Record<string, unknown>, key: string): string | undefined {
  const v = data[key];
  if (typeof v === "string" && v.trim().length > 0) return v.trim();
  return undefined;
}

function isValidEmail(email: string): boolean {
  const at = email.indexOf("@");
  if (at < 1 || at === email.length - 1) return false;
  const domain = email.slice(at + 1);
  return domain.includes(".");
}

export function generateIntakeId(): string {
  const hex = crypto.randomUUID().replace(/-/g, "").slice(0, 12);
  return `lead_${hex}`;
}

export function corsOrigins(raw: string | undefined): string[] {
  if (!raw) return [];
  return raw
    .split(",")
    .map((s) => s.trim())
    .filter(Boolean);
}
